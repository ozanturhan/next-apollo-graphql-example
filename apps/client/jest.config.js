module.exports = {
  coverageDirectory: '../../coverage/apps/client',
  testRegex: '(/__test__/.*|\\.(test|spec))\\.(ts|tsx|js|jsx)$',
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
  moduleNameMapper: {
    '@components': '<rootDir>/components',
    '@queries': '<rootDir>/graphql/queries',
    '\\.(css)$': '<rootDir>/__mocks__/styleMock.js',
    '\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$':
      '<rootDir>/__mocks__/fileMock.js',
  },
  transform: {
    '^.+\\.(js|jsx|ts|tsx)$': 'babel-jest',
  },
  setupFiles: ['./setupJest.ts'],
};
