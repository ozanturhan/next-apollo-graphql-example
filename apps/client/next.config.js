const withImages = require('next-images');
module.exports = withImages();

module.exports = {
  distDir: '../../dist/.next',
};
