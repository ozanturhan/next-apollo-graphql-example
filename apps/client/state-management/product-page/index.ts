export * from './actions/product-page.actions';
export * from './reducers/product-page.reducer';
export * from './states/product-page.state';
