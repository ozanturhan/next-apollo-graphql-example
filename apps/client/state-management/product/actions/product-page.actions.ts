import { Product } from '../../../interfaces/product';

export const FILL_PRODUCTS = 'FILL_PRODUCTS';
export const TOGGLE_PRODUCT_LIKE = 'TOGGLE_PRODUCT_LIKE';
export const CHANGE_FILTER = 'CHANGE_FILTER';

export class FillProductsAction {
  readonly type = FILL_PRODUCTS;
  constructor(public payload: Product[]) {}
}

export class ToggleLikeAction {
  readonly type = TOGGLE_PRODUCT_LIKE;
  constructor(public payload: number) {}
}

export class ChangeFilterAction {
  readonly type = CHANGE_FILTER;
  constructor(public payload: 'All' | 'Liked') {}
}

export type ProductActions = FillProductsAction | ToggleLikeAction | ChangeFilterAction;
