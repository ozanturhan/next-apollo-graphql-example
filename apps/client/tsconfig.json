{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "target": "es6",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noImplicitAny": false,
    "jsx": "preserve",
    "baseUrl": "./",
    "paths": {
      "@components": ["./components"],
      "@queries": ["./graphql/queries"]
    }
  },
  "include": ["packages/client/next-env.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["node_modules"]
}
