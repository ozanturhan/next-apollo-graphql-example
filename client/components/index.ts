export { Layout } from './Layout/Layout';
export { Wrapper } from './Wrapper/Wrapper';
export { Box } from './Box/Box';
export { BoxText } from './BoxText/BoxText';
export { LikeCount } from './LikeCount/LikeCount';
export { FilterButton } from './FilterButton/FilterButton';
export { LikeButton } from './LikeButton/LikeButton';
export { ProductPrice } from './ProductPrice/ProductPrice';
export { ProductTitle } from './ProductTitle/ProductTitle';
export { ProductDelivery } from './ProductDelivery/ProductDelivery';
export { ProductImage } from './ProductImage/ProductImage';
export { ProductCard } from './ProductCard/ProductCard';
export { ProductCardContainer } from './ProductCardContainer/ProductCardContainer';
export { ButtonGroup } from './ButtonGroup/ButtonGroup';
