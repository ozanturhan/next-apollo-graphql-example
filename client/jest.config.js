module.exports = {
  testPathIgnorePatterns: ['/.next/', '/node_modules/', '/tests/', '/coverage/'],
  testRegex: '(/__test__/.*|\\.(test|spec))\\.(ts|tsx|js|jsx)$',
  moduleFileExtensions: ['ts', 'tsx', 'js', 'json'],
  moduleNameMapper: {
    '@components': '<rootDir>/components',
    '@queries': '<rootDir>/queries',
    '\\.(css)$': '<rootDir>/__mocks__/styleMock.js',
    '\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$':
      '<rootDir>/__mocks__/fileMock.js',
  },
  transform: {
    '^.+\\.(js|jsx|ts|tsx)$': 'babel-jest',
  },
  transformIgnorePatterns: ['<rootDir>/node_modules/'],
  setupFiles: ['./setupJest.ts'],
};
