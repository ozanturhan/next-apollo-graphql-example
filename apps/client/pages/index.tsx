import React from 'react';
import {
  LayoutButtonGroup,
  LayoutMain,
  LayoutProductCard,
  LayoutWrapper,
  MoleculeFilterButton,
  MoleculeLikeCount,
  OrganismProductCard,
} from '@components';
import { ToggleLikeAction, useProductPageReducer } from '@state-management/product-page';
import { PRODUCTS_QUERY } from '@queries';
import { initializeApollo } from '@lib';
import Head from 'next/head';

function Home() {
  const { products, filter, likeCount, handleFilter, dispatch } = useProductPageReducer();

  return (
    <div>
      <Head>
        <title>Challenge</title>
        <meta name="Description" content="Challenge page." />
      </Head>
      <LayoutMain>
        <LayoutWrapper>
          <LayoutButtonGroup>
            <MoleculeLikeCount count={likeCount} />

            <MoleculeFilterButton disabled={likeCount === 0} onToggle={handleFilter} isToggle={filter === 'Liked'}>
              Beğendiklerim
            </MoleculeFilterButton>
          </LayoutButtonGroup>

          <LayoutProductCard title="İlginizi Çekebilecek Ürünler">
            {products?.map((item) => (
              <OrganismProductCard
                key={item.id}
                item={item}
                onToggleLike={() => dispatch(new ToggleLikeAction(item.id))}
              />
            ))}
          </LayoutProductCard>
        </LayoutWrapper>
      </LayoutMain>
    </div>
  );
}

export const getServerSideProps = async () => {
  const apolloClient = initializeApollo();

  await apolloClient.query({
    query: PRODUCTS_QUERY,
  });

  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
      revalidate: 1,
    },
  };
};

export default Home;
