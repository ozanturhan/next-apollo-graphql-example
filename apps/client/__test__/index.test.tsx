import React from 'react';
import Home, { getServerSideProps } from '../pages';
import { MockedProvider } from '@apollo/client/testing';
import { PRODUCTS_QUERY } from '@queries';
import { initializeApollo } from '@lib';
import { fireEvent, render, waitFor } from '@testing-library/react';
import * as productPageActions from '@state-management/product-page/actions/product-page.actions';
import { ToggleLikeAction } from '@state-management/product-page/actions/product-page.actions';

jest.mock('../lib/apollo-client');
const mockResult = {
  data: {
    products: [
      {
        id: 1,
        image: 'http://localhost:4001/image/256/iphone_11.jpg',
        isLiked: false,
        name: 'Apple iPhone 11 128 GB Pro Cep Telefonu Apple Türkiye Garantili',
        price: 1250,
        cargoType: 'FREE',
      },
      {
        id: 2,
        image: 'http://localhost:4001/image/256/iphone_11.jpg',
        isLiked: true,
        name: 'Apple iPhone 12 128 GB Pro Cep Telefonu Apple Türkiye Garantili',
        price: 1250,
        cargoType: 'FREE',
      },
    ],
  },
};

const mocks = [
  {
    request: {
      query: PRODUCTS_QUERY,
    },
    result: mockResult,
  },
];

let IndexPage = (
  <MockedProvider mocks={mocks} addTypename={false}>
    <Home />
  </MockedProvider>
);

describe('Index', () => {
  it('should be rendered correctly', async () => {
    // Arrange && Act
    const screen = render(IndexPage);

    await waitFor(() => screen.getAllByTestId('like-button'));

    // Assertion
    expect(screen.container).toMatchSnapshot();
  });

  it('should be filtered when clicking filter button', async () => {
    // Arrange
    const screen = render(IndexPage);

    await waitFor(() => screen.getAllByTestId('like-button'));

    fireEvent(
      screen.getByTestId('filter-button'),
      new MouseEvent('click', {
        bubbles: true,
      }),
    );

    // Assertion
    expect(screen.getAllByTestId('product')).toHaveLength(1);

    // Act
    fireEvent(
      screen.getByTestId('filter-button'),
      new MouseEvent('click', {
        bubbles: true,
      }),
    );

    // Assertion
    expect(screen.getAllByTestId('product')).toHaveLength(2);
  });

  it('like status should be changed when clicking like button', async () => {
    // Arrange
    const screen = render(IndexPage);

    const spyToggleLikeAction = jest
      .spyOn(productPageActions, 'ToggleLikeAction')
      .mockImplementation((payload: number) => {
        return (class {
          constructor(payload: number) {
            console.log(payload);
          }
        } as unknown) as ToggleLikeAction;
      });

    await waitFor(() => screen.getAllByTestId('like-button'));

    fireEvent(
      screen.getAllByTestId('like-button')[0],
      new MouseEvent('click', {
        bubbles: true,
      }),
    );

    // Assertion
    expect(spyToggleLikeAction).toHaveBeenCalledWith(1);
  });

  it('should get products form server', async () => {
    // Arrange
    const query = jest.fn();

    (initializeApollo as any).mockImplementation(() => ({
      query,
      cache: {
        extract: () => mockResult,
        restore: jest.fn(),
      },
    }));

    // Act
    const response = await getServerSideProps();

    // Assertion
    expect(response).toEqual(
      expect.objectContaining({
        props: {
          initialApolloState: mockResult,
          revalidate: 1,
        },
      }),
    );
  });
});
