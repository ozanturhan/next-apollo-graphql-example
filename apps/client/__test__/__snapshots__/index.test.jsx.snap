// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index should be rendered correctly 1`] = `
<div>
  <div
    class="Layout-zrzck-0 bgOuGU"
  >
    <div
      class="Wrapper-sc-1lkgs7-0 jjnCXr"
    >
      <div
        class="ButtonGroup-ruhkmw-0 fitOAb"
      >
        <div
          class="Box-sc-1hbv20g-0 hGtRHH"
        >
          <svg
            class="LikeCount__Heart-c1av5f-0 jNpQuz"
            viewBox="0 0 391.837 391.837"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M285.257 35.528c58.743.286 106.294 47.836 106.58 106.58 0 107.624-195.918 214.204-195.918 214.204S0 248.165 0 142.108c0-58.862 47.717-106.58 106.58-106.58a105.534 105.534 0 0 1 89.339 48.065 106.578 106.578 0 0 1 89.338-48.065z"
              fill="#d7443e"
            />
          </svg>
          <span
            class="BoxText-a8f6lv-0 fhwVnc"
          >
            1
             ürün
          </span>
        </div>
        <div
          class="Box-sc-1hbv20g-0 doRRdl"
        >
          <button
            class="FilterButton__Button-sc-9jzwea-0 gguiys"
          >
            Beğendiklerim
          </button>
        </div>
      </div>
      <div
        class="ProductCardContainer__Container-sc-1axx2t2-0 bBEipT"
      >
        <h2>
          İlginizi Çekebilecek Ürünler
        </h2>
        <div
          class="ProductCardContainer__GridContainer-sc-1axx2t2-1 ibgoXt"
        >
          <div
            class="ProductCard__Container-jr7vdt-0 ibiiel"
          >
            <button
              class="LikeButton__Button-nl0htv-0 dsbjUz"
            >
              <svg
                class="LikeButton__Heart-nl0htv-1 ABSrV"
                viewBox="0 0 391.837 391.837"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M285.257 35.528c58.743.286 106.294 47.836 106.58 106.58 0 107.624-195.918 214.204-195.918 214.204S0 248.165 0 142.108c0-58.862 47.717-106.58 106.58-106.58a105.534 105.534 0 0 1 89.339 48.065 106.578 106.578 0 0 1 89.338-48.065z"
                  fill="#d7443e"
                />
              </svg>
            </button>
            <a
              href="http://localhost:4001/image/256/iphone_11.jpg"
            >
              <span
                class="ProductImage__Container-jau957-0 eEAwFj"
              >
                <img
                  alt="http://localhost:4001/image/256/iphone_11.jpg"
                  src="http://localhost:4001/image/256/iphone_11.jpg"
                />
              </span>
              <span
                class="ProductTitle-sc-1tx9xcg-0 cIqNJk"
                data-testid="product-title"
              >
                Apple iPhone 11 128 GB Pro Cep Telefonu Apple Türkiye Garantili
              </span>
              <span
                class="ProductPrice-sc-17u85r7-0 kDOrHs"
              >
                1250
              </span>
              <span
                class="ProductDelivery__ProductDeliveryContainer-sc-1bjux91-0 jRsjew"
              >
                <svg
                  class="ProductDelivery__Icon-sc-1bjux91-1 ciTpiH"
                  viewBox="0 -104 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M365 3.016c44.39 10.3 61.371 36.257 82.66 71.984H365zM105.578 78c8.692 0 15 4.195 15 14 0 8.27-6.691 14.977-14.957 15H15c-8.285 0-15 6.719-15 15 0 8.285 6.715 15 15 15h135c8.363 0 15.059 6.71 15.059 15 0 8.285-6.715 15-15 15H15c-8.285 0-15 6.715-15 15s6.715 15 15 15h33v45c0 8.285 6.715 15 15 15h30.152c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46h152.304c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46H497c8.285 0 15-6.715 15-15v-90c0-44.012-46.422-46.934-46.465-47H350c-8.285 0-15-6.715-15-15V0H63c-8.285 0-15 6.715-15 15v33H30c-8.285 0-15 6.715-15 15s6.715 15 15 15zm328.367 148.055c10.739 10.738 10.739 28.156 0 38.894C416.672 282.223 387 269.934 387 245.5c0-24.43 29.672-36.719 46.945-19.445zm-265 0c10.739 10.738 10.739 28.156 0 38.894C151.672 282.223 122 269.934 122 245.5c0-24.43 29.672-36.719 46.945-19.445zm0 0"
                    fill="#1CB55E"
                    fill-rule="evenodd"
                  />
                </svg>
                 
                Ücretsiz
              </span>
            </a>
          </div>
          <div
            class="ProductCard__Container-jr7vdt-0 ibiiel"
          >
            <button
              class="LikeButton__Button-nl0htv-0 ggUova"
            >
              <svg
                class="LikeButton__Heart-nl0htv-1 ABSrV"
                viewBox="0 0 391.837 391.837"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M285.257 35.528c58.743.286 106.294 47.836 106.58 106.58 0 107.624-195.918 214.204-195.918 214.204S0 248.165 0 142.108c0-58.862 47.717-106.58 106.58-106.58a105.534 105.534 0 0 1 89.339 48.065 106.578 106.578 0 0 1 89.338-48.065z"
                  fill="#d7443e"
                />
              </svg>
            </button>
            <a
              href="http://localhost:4001/image/256/iphone_11.jpg"
            >
              <span
                class="ProductImage__Container-jau957-0 eEAwFj"
              >
                <img
                  alt="http://localhost:4001/image/256/iphone_11.jpg"
                  src="http://localhost:4001/image/256/iphone_11.jpg"
                />
              </span>
              <span
                class="ProductTitle-sc-1tx9xcg-0 cIqNJk"
                data-testid="product-title"
              >
                Apple iPhone 12 128 GB Pro Cep Telefonu Apple Türkiye Garantili
              </span>
              <span
                class="ProductPrice-sc-17u85r7-0 kDOrHs"
              >
                1250
              </span>
              <span
                class="ProductDelivery__ProductDeliveryContainer-sc-1bjux91-0 jRsjew"
              >
                <svg
                  class="ProductDelivery__Icon-sc-1bjux91-1 ciTpiH"
                  viewBox="0 -104 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M365 3.016c44.39 10.3 61.371 36.257 82.66 71.984H365zM105.578 78c8.692 0 15 4.195 15 14 0 8.27-6.691 14.977-14.957 15H15c-8.285 0-15 6.719-15 15 0 8.285 6.715 15 15 15h135c8.363 0 15.059 6.71 15.059 15 0 8.285-6.715 15-15 15H15c-8.285 0-15 6.715-15 15s6.715 15 15 15h33v45c0 8.285 6.715 15 15 15h30.152c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46h152.304c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46H497c8.285 0 15-6.715 15-15v-90c0-44.012-46.422-46.934-46.465-47H350c-8.285 0-15-6.715-15-15V0H63c-8.285 0-15 6.715-15 15v33H30c-8.285 0-15 6.715-15 15s6.715 15 15 15zm328.367 148.055c10.739 10.738 10.739 28.156 0 38.894C416.672 282.223 387 269.934 387 245.5c0-24.43 29.672-36.719 46.945-19.445zm-265 0c10.739 10.738 10.739 28.156 0 38.894C151.672 282.223 122 269.934 122 245.5c0-24.43 29.672-36.719 46.945-19.445zm0 0"
                    fill="#1CB55E"
                    fill-rule="evenodd"
                  />
                </svg>
                 
                Ücretsiz
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
